_posts/2024-08-06-how-to-talk-about-CHERI.markdown

```diff
@@ -0,0 +1,185 @@
+For example, there's no Rust wrapper that you can write that prevents a C library (on non-CHERI hardware) from mutating an immutable object, or from capturing a borrowed reference.
+
+But what if the hardware could help?
+What if the Rust compiler could tell some combination of the hardware and the language-agnostic run-time system that a pointer must not be captured, or may be used with load instructions but not store?
+
+## So what is this chariot thing you kids are talking about?
+
+CHERIoT allows a compiler to enforce a fairly rich set of properties against other bits of a program.
+It can guarantee that, when invoking untrusted code, the untrusted code cannot:
+
+ - Access objects unless passed pointers to them.
+ - Access an object (or the memory that was formerly used for the object) after the object has been freed.
+ - Hold a pointer to an object with automatic storage duration (an 'on-stack' object) after the end of the call in which it was created.
+ - Hold a temporarily delegated pointer beyond a single call.
+ - Modify an object passed via immutable reference.
+ - Modify any object reachable from an object that is passed as a deeply immutable reference.
+ - Tamper with (or access) an object passed via opaque reference.
+
+Some of these are baseline guarantees from CHERI, some are built atop those with additional hardware extensions, and others are built atop them with some language-agnostic software.
+Let's go through them one at a time.
+
+Running code cannot access objects unless passed pointers to them.
+This is a basic property of CHERI but also the absolute minimum foundation for *any* form of memory safety.
+If malicious or buggy code can materialise arbitrary pointers out of thin air and use them, it can bypass any of the other rules.
+In any CHERI system, all instructions that accesses memory (loads, stores, jumps, and so on) must take a capability as one of its operands.
+A capability (in general) is an unforgeable token of authority that allows you to do some action.
+A CHERI capability is a machine word that is interpreted by the CPU and which can be used as a hardware type that *represents the language notion of a pointer*.
+CHERI capabilities were designed to allow everything that correct C does with pointers, which is a superset of what most safer languages permit.
+
+That capability is protected both in registers and memory by a tag bit.
+The tag bit is an attestation from the hardware that there is a valid chain of provenance from the root capabilities that authorise everything down via a sequence of subsetting operations that end up with the capability that you hold.
+On CHERIoT, for example, if you have a capability to a heap object, this is telling you that the loader restricted a read-write capability to cover the address space that's used for the heap, then the allocator subset it to give you a pointer to an object.
```

Comment:
This feels like a complex sentence for trying to get the point across - you don't have to explain all the details in this post, just enough of the key point to show why its important.  Maybe something like:  

On CHERIoT, for example if you have a capability to a heap object with a valid tag, you can trust that it was created from another capability that had access to to the heap address space.